platform/security/sts/security-sts-server/src/main/java/ddf/security/sts/SAMLSubjectProvider.java

```diff
@@ -0,0 +1,200 @@
+import org.apache.wss4j.common.saml.bean.KeyInfoBean;
+import org.apache.wss4j.common.saml.bean.SubjectBean;
+import org.apache.wss4j.common.saml.builder.SAML2Constants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.Document;
+
+import ddf.security.SecurityConstants;
+
+/**
+ * Custom SubjectProvider that sets the NameIDFormat
+ * based on the type of Principal.
+ */
+public class SAMLSubjectProvider extends DefaultSubjectProvider {
+
+    /**
+     * Log4j Logger
+     */
+    private static final Logger LOGGER = LoggerFactory.getLogger(SecurityConstants.SECURITY_LOGGER);
+
+    @Override
+    /**
+     * Get a SubjectBean object.
+     */ public SubjectBean getSubject(TokenProviderParameters providerParameters, Document doc,
+            byte[] secret) {
+        String subjectNameQualifier = "http://cxf.apache.org/sts";
+        String subjectNameIDFormat;
+
+        TokenRequirements tokenRequirements = providerParameters.getTokenRequirements();
+        KeyRequirements keyRequirements = providerParameters.getKeyRequirements();
+        STSPropertiesMBean stsProperties = providerParameters.getStsProperties();
+
+        String tokenType = tokenRequirements.getTokenType();
+        String keyType = keyRequirements.getKeyType();
+        String confirmationMethod = getSubjectConfirmationMethod(tokenType, keyType);
+
+        Principal principal = null;
+        ReceivedToken receivedToken;
+        //TokenValidator in IssueOperation has validated the ReceivedToken
+        //if validation was successful, the principal was set in ReceivedToken
+        if (providerParameters.getTokenRequirements().getOnBehalfOf() != null) {
+            receivedToken = providerParameters.getTokenRequirements().getOnBehalfOf();
+            if (receivedToken.getState().equals(ReceivedToken.STATE.VALID)) {
+                principal = receivedToken.getPrincipal();
+            }
+        } else if (providerParameters.getTokenRequirements().getActAs() != null) {
+            receivedToken = providerParameters.getTokenRequirements().getActAs();
+            if (receivedToken.getState().equals(ReceivedToken.STATE.VALID)) {
+                principal = receivedToken.getPrincipal();
+            }
+        } else if (providerParameters.getTokenRequirements().getValidateTarget() != null) {
+            receivedToken = providerParameters.getTokenRequirements().getValidateTarget();
+            if (receivedToken.getState().equals(ReceivedToken.STATE.VALID)) {
+                principal = receivedToken.getPrincipal();
+            }
+        } else {
+            principal = providerParameters.getPrincipal();
+        }
+
+        if (principal == null) {
+            LOGGER.error("Error in getting principal");
+            throw new STSException("Error in getting principal", STSException.REQUEST_FAILED);
+        } else {
+            if (principal instanceof UsernameTokenPrincipal) {
```

Comment:
I think we could. We'd need to make a couple changes though to preserve their current functionality. In other words, if someone has set the subjectNameIDFormat field to some value, then that value should be respected and it should always return that format, regardless of what the Principal is. However, if that field hasn't been set (i.e. it is still unspecified), then we would do this else block.